relay-conventions/build/name.rs

```diff
@@ -0,0 +1,151 @@
+///
+/// In other words, each template list must have a fallback name for the case
+/// where all names are missing. These invariants are enforced by a test in `sentry-conventions`.
+pub fn name_file_output(names: impl Iterator<Item = Name>) -> TokenStream {
+    let operations = names.flat_map(|name| name.operations);
+
+    let match_arms = TokenStream::from_iter(operations.map(|Operation { ops, templates }| {
+        // Each template stanza handles a single template: if all referenced attributes are present,
+        // the name is constructed via that template and returned.
+        let template_stanzas = templates.into_iter().flat_map(|template| {
+            let Some(parts) = parse_template_into_parts(&template) else {
+                println!(
+                    "sentry_conventions contained unparseable template \"{}\"",
+                    template
+                );
+                return None;
+            };
+
+            let first_part = parts.first()?;
+            let referenced_attribute_names: Vec<&str> = parts
+                .iter()
+                .filter_map(|part| match part {
+                    TemplatePart::Literal(_) => None,
+                    TemplatePart::Attribute(attr) => Some(*attr),
+                })
+                .collect();
+
+            if referenced_attribute_names.is_empty() {
+                // There are no dynamic parts to this name - it consists of a single literal.
+                let static_name = match first_part {
+                    TemplatePart::Literal(s) => *s,
+                    _ => unreachable!(),
+                };
+                // Because of the enforced invariant that the single literal template is always
+                // last, no explicit return is needed. (Otherwise clippy complains.)
+                return Some(quote! {
+                    #static_name.to_owned()
+                });
+            }
+
+            // Ensure that all referenced attributes are present in `attributes` before constructing
+            // a name from them.
+            let if_clauses = referenced_attribute_names.iter().map(|attribute_name| {
+                quote! {
+                    attributes.get_value(#attribute_name).and_then(|value| value.as_str()).is_some()
+                }
+            });
+
+            // Append each part to the String variable `name`.
+            let append_parts_to_string =
+                TokenStream::from_iter(parts.iter().map(|part| match part {
+                    TemplatePart::Literal(s) => {
+                        // This feels like a micro-optimization, but clippy complains without it.
+                        if s.chars().count() == 1 {
+                            let c = s.chars().next();
+                            quote! { name.push(#c); }
+                        } else {
+                            quote! { name.push_str(#s); }
+                        }
```

Comment:
Could also slap an `allow` on this. Either way should be fine.
```suggestion
                        quote! {
                            #[allow(clippy::single_char_add_str)]
                            name.push_str(#s);
                        }
```

Reply:
Cool. Used an `allow` here and anywhere else it complicated the codegen.